### DISCUSSION TITLE
Modal hide function is not working in bootstrap(v5)

### DISCUSSION THREAD

**Discussion by devbluemoons** [twbs/bootstrap#32347] [Q&A] [answered]
At first, thanks a lot to provide the bootstrap 5 version

I found not working `hide()` function in Modal

Please, check this issue
```html
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
    <head>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
        <meta http-equiv="Content-Style-Type" content="text/css" />
        <title></title>
        <link
            rel="stylesheet"
            href="https://stackpath.bootstrapcdn.com/bootstrap/5.0.0-alpha2/css/bootstrap.min.css"
            integrity="sha384-DhY6onE6f3zzKbjUPRc2hOzGAdEf4/Dz+WJwBvEYL/lkkIsI3ihufq9hk9K4lVoK"
            crossorigin="anonymous"
        />
        <script
            src="https://stackpath.bootstrapcdn.com/bootstrap/5.0.0-alpha2/js/bootstrap.bundle.min.js"
            integrity="sha384-BOsAfwzjNJHrJ8cZidOg56tcQWfp6y72vEJ8xQ9w6Quywb24iOsW913URv1IS4GD"
            crossorigin="anonymous"
        ></script>
        <style type="text/css"></style>
    </head>
    <body>
        <button id="btnShow">Modal</button>
        <div class="modal fade" tabindex="-1" id="testModal">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <h5 class="modal-title">Modal title</h5>
                        <button type="button" class="btn-close" data-dismiss="modal" aria-label="Close"></button>
                    </div>
                    <div class="modal-body">
                        <p>Modal body text goes here.</p>
                    </div>
                    <div class="modal-footer">
                        <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
                        <button type="button" class="btn btn-primary" id="btnSave">Save</button>
                    </div>
                </div>
            </div>
        </div>
    </body>
    <script type="text/javascript">
        window.onload = function () {
            document.getElementById("btnShow").addEventListener("click", function () {
                const container = document.getElementById("testModal");
                const modal = new bootstrap.Modal(container);
                modal.show();
            });
            document.getElementById("btnSave").addEventListener("click", function () {
                const container = document.getElementById("testModal");
                const modal = new bootstrap.Modal(container);
                modal.hide();
            });
        };
    </script>
</html>
```

**Comment by XhmikosR** [comment, marked as answer]
You are doing it wrong; you need to use `hide()` on the same modal instance. Also, you don't need that obsolete HTML4 stuff and you should always [validate](https://validator.w3.org/) your HTML.

```html
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Modal test</title>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-alpha3/dist/css/bootstrap.min.css">
    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-alpha3/dist/js/bootstrap.bundle.min.js"></script>
  </head>
  <body>
    <button id="btnShow" type="button" class="btn btn-primary">Modal</button>
    <div class="modal fade" tabindex="-1" id="testModal">
      <div class="modal-dialog">
        <div class="modal-content">
          <div class="modal-header">
            <h5 class="modal-title">Modal title</h5>
            <button type="button" class="btn-close" data-dismiss="modal" aria-label="Close"></button>
          </div>
          <div class="modal-body">
            <p>Modal body text goes here.</p>
          </div>
          <div class="modal-footer">
            <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
            <button type="button" class="btn btn-primary" id="btnSave">Save</button>
          </div>
        </div>
      </div>
    </div>
    <script>
      const container = document.getElementById("testModal");
      const modal = new bootstrap.Modal(container);

      document.getElementById("btnShow").addEventListener("click", function () {
        modal.show();
      });
      document.getElementById("btnSave").addEventListener("click", function () {
        modal.hide();
      });
    </script>
  </body>
</html>
```

**Reply by devbluemoons** [reply]
Thank you very much for help^^
That is not an issue, It's just my mistake!

**Reply by webdevluc** [reply]
Thank you very much XhmikosR your solution helped me a lot with my modal hiding situation. Much appreciated!

**Reply by ghost** [reply]
Hi,
I followed you sample, it works with show/hide,
but it alos close the dialog, when I click outside the dialog, how can I keep the dialog open and only close it by click the button?
Thanks
Regards,
Gordon

**Reply by insidert** [reply]
https://getbootstrap.com/docs/5.0/components/modal/#static-backdrop

Check this on docs.

**Reply by ghost** [reply]
Thanks, 
I copied the dialog, use the modal.show(); to get it show, and it still hide by click outside the dialog.
I need use javacript to get it show as I need dosomething before it shown.

**Reply by insidert** [reply]
Add this attribute data-bs-backdrop="static" 

```html
<div class="modal fade" id="staticBackdrop" data-bs-backdrop="static" data-bs-keyboard="false" tabindex="-1" aria-labelledby="staticBackdropLabel" aria-hidden="true">
```

**Reply by ghost** [reply]
Thank you,
It is fixed, I have noticed that I am not using the 5.0, instead it is 4.3.1.
I have upgraded it to 5.0 and it is working now.

**Reply by TechDan09** [reply]
Thanks @XhmikosR

**Reply by angelokezimana** [reply]
> You are doing it wrong; you need to use `hide()` on the same modal instance. Also, you don't need that obsolete HTML4 stuff and you should always [validate](https://validator.w3.org/) your HTML.
> 
> ```
> <!doctype html>
> <html>
>   <head>
>     <meta charset="utf-8">
>     <title>Modal test</title>
>     <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-alpha3/dist/css/bootstrap.min.css">
>     <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-alpha3/dist/js/bootstrap.bundle.min.js"></script>
>   </head>
>   <body>
>     <button id="btnShow" type="button" class="btn btn-primary">Modal</button>
>     <div class="modal fade" tabindex="-1" id="testModal">
>       <div class="modal-dialog">
>         <div class="modal-content">
>           <div class="modal-header">
>             <h5 class="modal-title">Modal title</h5>
>             <button type="button" class="btn-close" data-dismiss="modal" aria-label="Close"></button>
>           </div>
>           <div class="modal-body">
>             <p>Modal body text goes here.</p>
>           </div>
>           <div class="modal-footer">
>             <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
>             <button type="button" class="btn btn-primary" id="btnSave">Save</button>
>           </div>
>         </div>
>       </div>
>     </div>
>     <script>
>       const container = document.getElementById("testModal");
>       const modal = new bootstrap.Modal(container);
> 
>       document.getElementById("btnShow").addEventListener("click", function () {
>         modal.show();
>       });
>       document.getElementById("btnSave").addEventListener("click", function () {
>         modal.hide();
>       });
>     </script>
>   </body>
> </html>
> ```

Thanks very much @XhmikosR, your comment helped me as well.

**Reply by ParmarMansi** [reply]
I'm not coder like you but i can suggest you the [courses](https://teacherdada.com/course/it-amp-development/learn-javascript-basics-tutorial-for-beginners-in-marathi-188) that can help you to clear basics. It might not help you but you can suggest some one who are going to start his/her career in software industry.

**Comment by insidert** [comment]
@XhmikosR comment helped me. 
For people writing Vue, and using JS to show and hide the modal, you need to add the modal to the data object.

```js
data: {
 someModal: ""
},

// other code
someMethod() {
 this.someModal = new bootstrap.Modal(document.getElementById('modalId'));
 this.someModal.show();
 this.someModal.hide();
}
```

**Reply by josephmbabu** [reply]
I'm working on a side project wherein I'm using Bootstrap 5, Vue 3, and Vue's Composition API. I have written a utility method, **_toggleModal()_**, to handle the showing and hiding of the modal.  **_modal.hide()_** is not responding accordingly when the  _save_ button  in the template is clicked; it only closes the modal when **_data-dismiss="modal"_** is added to the button element, and even then the modal is closed before the async call to the API is completed. I have included a stripped-out js from the component.  Any assistance on what I'm missing regarding the modal implementation with Vuejs will be much appreciated. Thanks.

export default {
    name: "SomeComponent",
    setup() {
        const isOpen =  ref(false);
        const modal = ref("");
        
        const toggleModal = () => {
            modal.value = new Modal(document.getElementById("modalId"));
            isOpen.value = !isOpen.value;
            if (isOpen.value) {
                modal.value.show();
            } else {
                modal.value.hide();
            }
        };

        const handleNew = () => {
            toggleModal();
        };

        return {
            isOpen,
            toggleModal,
            handleNew
        };
}

**Reply by codemains** [reply]
Much thanks for the help @insidert . Totally worked for me :)

**Reply by insidert** [reply]
> I'm working on a side project wherein I'm using Bootstrap 5, Vue 3, and Vue's Composition API. I have written a utility method, **_toggleModal()_**, to handle the showing and hiding of the modal. **_modal.hide()_** is not responding accordingly when the _save_ button in the template is clicked; it only closes the modal when **_data-dismiss="modal"_** is added to the button element, and even then the modal is closed before the async call to the API is completed. I have included a stripped-out js from the component. Any assistance on what I'm missing regarding the modal implementation with Vuejs will be much appreciated. Thanks.
> 
> export default {
> name: "SomeComponent",
> setup() {
> const isOpen = ref(false);
> const modal = ref("");
> 
> ```
>     const toggleModal = () => {
>         modal.value = new Modal(document.getElementById("modalId"));
>         isOpen.value = !isOpen.value;
>         if (isOpen.value) {
>             modal.value.show();
>         } else {
>             modal.value.hide();
>         }
>     };
> 
>     const handleNew = () => {
>         toggleModal();
>     };
> 
>     return {
>         isOpen,
>         toggleModal,
>         handleNew
>     };
> ```
> 
> }

I think, you should not assign the modal inside the function. It should be assigned like in created() method in V2. So, V3, assign the modal outside the toggling function. Since v3, created methods are inside the setup. Sorry for the delay in response.

**Reply by marsjaninzmarsa** [reply]
I've came about this issue and after few hours of debugging and looking for solutions I must say: docs section for that component is very misleading, and only place where we can came across the correct approach, is one, single, enigmatic paragraph in the [Getting started](https://getbootstrap.com/docs/5.0/getting-started/javascript/#programmatic-api) guide, very easy to miss, and without reading components source code, very hard to understand it's point.

Tl;dr: instead of creating a new instance of Modal object each time (each one with its own, independent state) for each action, try reference directly to the object assigned to your selected DOM node the same way as default controller for default buttons does, eg:
```
myModal = Modal.getOrCreateInstance(document.querySelector('.modal-css-selector'));
myModal.show();
myModal.hide();
```

**Comment by MJORR** [comment]
See documentation: https://getbootstrap.com/docs/5.0/components/modal/

New Way
` <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>`

Old Way
`<button type="button" class="btn-close" data-dismiss="modal" aria-label="Close"></button>`

**Reply by yarrarajendra** [reply]
Thank you very much, Its works for me.

**Comment by MacchiFilippo** [comment]
Well I have a similar problem with modal.hide.

If I use this code:
var waitingDialog = waitingDialog || (function ($) {
    'use strict';    
    var $modal = new bootstrap.Modal(document.getElementById('mdialog'));
    
    return {
        show: function () {
            $modal.show();
        },        
        hide: function () {
            $modal.hide();
        }
    };

})(jQuery);

I can show correctly the modal element but the hide function doesn't work and I don't understand why.

Anyone have an idea?

**Reply by GeoSot** [reply]
https://github.com/twbs/bootstrap/issues/34952#issuecomment-917937829

**Comment by JeffLutzenberger** [comment]
if your `show` and `hide` functions live in different scopes and you don't want to hold onto the modal object this seems to work: 

`bootstrap.Modal.getOrCreateInstance(document.getElementById('your-element-id')).hide()`

**Reply by httpanand** [reply]
can anyone please help me with this ? https://github.com/twbs/bootstrap/discussions/35346

**Reply by aghabalaguluzade** [reply]
var myModalEl = document.querySelector('#scheduleMeetingModal');
var myModal = bootstrap.Modal.getOrCreateInstance(myModalEl);
myModal.hide();

**Reply by bigjam** [reply]
same problem here !
Variable are not null but hide doesn't work.

**Reply by ghost** [reply]
Not work in jQuery ajax callback ...
